src/bun.js/api/bun/h2_frame_parser.zig

```diff
@@ -0,0 +1,2440 @@
+            }
+        }
+
+        if (handlers.onWrite == .zero) {
+            exception.* = JSC.toInvalidArguments("Expected at least \"write\" callback", .{}, globalObject).asObjectRef();
+            return null;
+        }
+
+        if (opts.getTruthy(globalObject, "binaryType")) |binary_type_value| {
+            if (!binary_type_value.isString()) {
+                exception.* = JSC.toInvalidArguments("Expected \"binaryType\" to be a string", .{}, globalObject).asObjectRef();
+                return null;
+            }
+
+            handlers.binary_type = BinaryType.fromJSValue(globalObject, binary_type_value) orelse {
+                exception.* = JSC.toInvalidArguments("Expected 'binaryType' to be 'arraybuffer', 'uint8array', 'buffer'", .{}, globalObject).asObjectRef();
+                return null;
+            };
+        }
+
+        return handlers;
+    }
+
+    pub fn unprotect(this: *Handlers) void {
+        this.onError.unprotect();
+        this.onGoAway.unprotect();
+        this.onWrite.unprotect();
+        this.onStreamError.unprotect();
+        this.onStreamStart.unprotect();
+        this.onStreamHeaders.unprotect();
+        this.onStreamEnd.unprotect();
+        this.onStreamData.unprotect();
+        this.onStreamError.unprotect();
+        this.onLocalSettings.unprotect();
+        this.onRemoteSettings.unprotect();
+        this.onWantTrailers.unprotect();
+        this.onPing.unprotect();
+        this.onEnd.unprotect();
+        this.onFrameError.unprotect();
+        this.onAborted.unprotect();
+    }
+
+    pub fn clear(this: *Handlers) void {
+        this.onError = .zero;
+        this.onWrite = .zero;
+        this.onStreamError = .zero;
+        this.onStreamStart = .zero;
+        this.onStreamHeaders = .zero;
+        this.onStreamEnd = .zero;
+        this.onStreamData = .zero;
+        this.onStreamError = .zero;
+        this.onLocalSettings = .zero;
+        this.onRemoteSettings = .zero;
+        this.onWantTrailers = .zero;
+        this.onPing = .zero;
+        this.onEnd = .zero;
+        this.onGoAway = .zero;
+        this.onFrameError = .zero;
+        this.onAborted = .zero;
+    }
+
+    pub fn protect(this: *Handlers) void {
+        this.onError.protect();
+        this.onWrite.protect();
+        this.onStreamError.protect();
+        this.onStreamStart.protect();
+        this.onStreamHeaders.protect();
+        this.onStreamEnd.protect();
+        this.onStreamData.protect();
+        this.onStreamError.protect();
+        this.onLocalSettings.protect();
+        this.onRemoteSettings.protect();
+        this.onWantTrailers.protect();
+        this.onEnd.protect();
+        this.onGoAway.protect();
+        this.onFrameError.protect();
+        this.onAborted.protect();
+    }
+};
+
+pub const H2FrameParser = struct {
+    pub const log = Output.scoped(.H2FrameParser, false);
+    pub usingnamespace JSC.Codegen.JSH2FrameParser;
+
+    strong_ctx: JSC.Strong = .{},
```

Comment:
Do you have any tests that check we aren't leaking the context?

Reply:
will add leak tests + more frame error/limits tests